packages/flutter/lib/src/widgets/selectable_region.dart

```diff
@@ -2896,3 +3023,396 @@ typedef SelectableRegionContextMenuBuilder = Widget Function(
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _selectableRegionScope?.removeListener(_handleOnSelectableRegionChanged);
+    _selectableRegionScope = _SelectableRegionScope.maybeOf(context);
+    _selectableRegionScope?.addListener(_handleOnSelectableRegionChanged);
+  }
+
+  @override
+  void didUpdateWidget(SelectionListener oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.controller != widget.controller) {
+      _selectionDelegate.setController(widget.controller);
+    }
+  }
+
+  @override
+  void dispose() {
+    _selectionDelegate.dispose();
+    super.dispose();
+  }
+
+  void _handleOnSelectableRegionChanged() {
+    if (_selectableRegionScope != null) {
+      _selectionDelegate._dispatchStatusEvent(_selectableRegionScope!.selectionListenerStatus);
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return SelectionContainer(
+      delegate: _selectionDelegate,
+      child: widget.child,
+    );
+  }
+}
+
+class _SelectionListenerDelegate extends _SelectableRegionContainerDelegate {
+  _SelectionListenerDelegate({
+    required SelectionListenerController controller,
+  }) : _controller = controller {
+    _controller._selectionDelegate = this;
+  }
+
+  SelectionListenerController _controller;
+  void setController(SelectionListenerController newController) {
+    _controller = newController;
+    _controller._selectionDelegate = this;
+  }
+
+  SelectionGeometry? _lastFinalizedSelectionGeometry;
+  SelectionGeometry? _previousSelectionGeometry;
+
+  void _dispatchStatusEvent(SelectionListenerStatus status) {
+    if (_previousSelectionGeometry != value) {
+      _controller.notifyListeners();
+      _controller.notifyStatusListeners(status);
+    } else if (status == SelectionListenerStatus.finalized && _lastFinalizedSelectionGeometry != value) {
+      _controller.notifyListeners();
+      _controller.notifyStatusListeners(status);
+      _lastFinalizedSelectionGeometry = value;
+    }
+    _previousSelectionGeometry = value;
+  }
+
+  @override
+  void dispose() {
+    _controller._unregisterSelectionListenerDelegate();
+    super.dispose();
+  }
+}
+
+/// Notifies listeners when the selection has been changed or finalized.
+final class SelectionListenerController extends ChangeNotifier {
+  _SelectionListenerDelegate? _selectionDelegate;
+
+  /// The computed selection range of the owning [SelectionListener]s subtree.
+  SelectedContentRange get value => _selectionDelegate?.getSelection() ?? (throw Exception('Selection client has not been registered to this controller.'));
+
+  /// The status that indicates whether there is a selection and whether the
+  /// selection is collapsed.
+  SelectionStatus get selectionStatus => _selectionDelegate?.value.status ?? (throw Exception('Selection client has not been registered to this controller.'));
+
+  void _unregisterSelectionListenerDelegate() {
+    _selectionDelegate = null;
+  }
+
+  // From ChangeNotifier.
```

Comment:
You're already extending `ChangeNotifier`, so why are these needed?

Reply:
These are for the status listener.

Reply:
Would it be possible to create a dedicated ChangeNotifier for status internally?

Reply:
If we did that I think we would still need this code from `ChangeNotifier` right? Or did you mean make listening to the status completely private, i.e. no `addStatusListener`, and make `SelectionListenerStatus` accessible as a getter through `SelectionListenerController` (renamed to `SelectionListenerNotifier`).

```dart
.addListener(() {
  mySelectionListenerStatus = _selectionListenerNotifier.selectionListenerStatus;
})
```

Reply:
I was just concerned about the code duplication. 

Is it possible to (lazily maybe) create a ChangeNotifier instance internally in this class to store the status listeners instead of implementing the logic for adding / removing status listeners yourself?

Reply:
I think this might not be possible since `ChangeNotifier` only accepts `VoidCallback`s as listeners, so we have to manage our own listener list if we want something custom. Unless you had something else in mind? We can get rid of the dupe code by going with this approach https://github.com/flutter/flutter/pull/154202#discussion_r1815498811 .

Reply:
This amount of code duplication should be avoided. Another option is to use `ObserverList`. 

> .addListener(() { mySelectionListenerStatus = _selectionListenerNotifier.selectionListenerStatus; })

This feels too verbose as a public API.

Reply:
Thanks for the suggestion, went with `ObserverList`.